File: Classes/Models/SearchXMLParser.m
//
//  SearchXMLParser.m
//  iSub
//
//  Created by bbaron on 10/21/10.
//  Copyright 2010 Ben Baron. All rights reserved.
//

#import "SearchXMLParser.h"
#import "iSubAppDelegate.h"
#import "ISMSArtist.h"
#import "ISMSAlbum.h"
#import "ISMSSong+DAO.h"

@interface SearchXMLParser() {
    __strong NSMutableArray<ISMSArtist*> *_listOfArtists;
    __strong NSMutableArray<ISMSAlbum*> *_listOfAlbums;
    __strong NSMutableArray<ISMSSong*> *_listOfSongs;
}
@end

@implementation SearchXMLParser

- (NSArray<ISMSArtist*> *)listOfArtists { return _listOfArtists; }
- (NSArray<ISMSAlbum*> *)listOfAlbums { return _listOfAlbums; }
- (NSArray<ISMSSong*> *)listOfSongs { return _listOfSongs; }

- (instancetype)init  {
	if (self = [super init]) {
		_listOfArtists = [[NSMutableArray alloc] init];
		_listOfAlbums = [[NSMutableArray alloc] init];
		_listOfSongs = [[NSMutableArray alloc] init];
	}

	return self;
}

- (void)parser:(NSXMLParser *)parser parseErrorOccurred:(NSError *)parseError {
    // TODO: uncomment this
	/*UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Subsonic Error" message:@"There was an error parsing the XML response. Subsonic may have had an error performing the search." delegate:appDelegateS cancelButtonTitle:@"Ok" otherButtonTitles:@"Settings", nil];
	alert.tag = 1;
	[alert show];*/
}


- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qualifiedName attributes:(NSDictionary *)attributeDict {
	if ([elementName isEqualToString:@"match"] || [elementName isEqualToString:@"song"]) {
		if (![[attributeDict objectForKey:@"isVideo"] isEqualToString:@"true"]) {
			ISMSSong *song = [[ISMSSong alloc] initWithAttributeDict:attributeDict];
			if (song.path) {
				[_listOfSongs addObject:song];
            }
		}
	} else if ([elementName isEqualToString:@"album"]) {
		[_listOfAlbums addObject:[[ISMSAlbum alloc] initWithAttributeDict:attributeDict]];
	} else if ([elementName isEqualToString:@"artist"]) {
		[_listOfArtists addObject:[[ISMSArtist alloc] initWithAttributeDict:attributeDict]];
	}
}

@end

File: Classes/Models/SearchXMLParser.swift
//
//  SearchXMLParser.swift
//  iSub
//
//  Created by Benjamin Baron on 12/22/20.
//  Copyright © 2020 Ben Baron. All rights reserved.
//

import Foundation
import CocoaLumberjackSwift
import Resolver

final class SearchXMLParser {
    @Injected private var store: Store
        
    let serverId: Int
    
    private(set) var folderArtists = [FolderArtist]()
    private(set) var folderAlbums = [FolderAlbum]()
    private(set) var songs = [Song]()
    
    init(serverId: Int, data: Data) {
        self.serverId = serverId
        let root = RXMLElement(fromXMLData: data)
        if !root.isValid {
            // TODO: Handle this error in the UI
            DDLogError("[SearchXMLParser] Error parsing search results: \(APIError.responseNotXML)")
        } else {
            if let error = root.child("error"), error.isValid {
                // TODO: Handle this error in the UI
                DDLogError("[SearchXMLParser] Subsonic error: \(SubsonicError(element: error))")
            } else {
                let isNewSearchSupported = store.server(id: serverId)?.isNewSearchSupported ?? false
                if isNewSearchSupported {
                    root.iterate("searchResult2.artist") { element, _ in
                        self.folderArtists.append(FolderArtist(serverId: serverId, element: element))
                    }
                    root.iterate("searchResult2.album") { element, _ in
                        self.folderAlbums.append(FolderAlbum(serverId: serverId, element: element))
                    }
                    root.iterate("searchResult2.song") { element, _ in
                        let isVideo = element.attribute("isVideo")
                        if isVideo != "true" {
                            let song = Song(serverId: serverId, element: element)
                            if song.path != "" {
                                self.songs.append(song)
                            }
                        }
                    }
                } else {
                    root.iterate("searchResult.match") { element, _ in
                        let isVideo = element.attribute("isVideo")
                        if isVideo != "true" {
                            let song = Song(serverId: serverId, element: element)
                            if song.path != "" {
                                self.songs.append(song)
                            }
                        }
                    }
                }
            }
        }
    }
}
